File: source/smc/generator/csharp/CSharpCodeGenerators/CSharpCodeGenerator.java
package smc.generator.csharp.CSharpCodeGenerators;

import smc.generator.csharp.SMCSharpGenerator;
import smc.fsmrep.State;

public abstract class CSharpCodeGenerator
{
    public abstract String generateCode(SMCSharpGenerator gen);

    public  String printSeparator( int level )
    {
        return  "";
    }

    public String createStateFieldName( State s )
    {
        StringBuffer buff = new StringBuffer(s.getName());
		buff.append("State");
        if( buff.length() > 0 )
            buff.setCharAt(0, Character.toLowerCase( buff.charAt(0) ));
        return( buff.toString() );
    }

    public String createMethodName( State s )
    {
        StringBuffer buff = new StringBuffer( s.getName() );
        if( buff.length() > 0 )
            buff.setCharAt(0, Character.toUpperCase( buff.charAt(0) ));
        return( buff.toString() );
    }

    public String createMethodName( String event )
    {
        StringBuffer buff = new StringBuffer( event );
        if( buff.length() > 0 )
            buff.setCharAt(0, Character.toUpperCase( buff.charAt(0) ));
        return( buff.toString() );
    }
}

File: source/smc/generator/csharp/CSharpCodeGenerators/CSharpCodeGenerator.cs
namespace SMC.Generator.CSharp.CSharpCodeGenerators
{
    using System;
    using System.Collections.Generic;
    using System.Text;

    using SMC.FsmRep;
    using SMC.Generator.CSharp;

    public abstract class CSharpCodeGenerator
    {
        public abstract string GenerateCode(SMCSharpGenerator gen);

        public string PrintSeparator(int level) => string.Empty;

        public string ClassNameFor(State s) => $"{CreateMethodName(s)}State";

        public string CreateMethodName(State s) => CreateMethodName(s.Name);

        public string CreateMethodName(string _event)
        {
            var length = _event.Length;
            var buff = new StringBuilder(length);
            if (length > 0)
            {
                buff.Append(char.ToUpperInvariant(_event[0]));
                if (length > 1)
                {
                    buff.Append(_event.Substring(1));
                }
            }

            return buff.ToString();
        }

        public static StringBuilder AddFromGenerators(SMCSharpGenerator gen, Func<IEnumerable<CSharpCodeGenerator>> generatorFactory)
            => AddFromGenerators(gen, generatorFactory, new StringBuilder());

        public static StringBuilder AddFromGenerators(SMCSharpGenerator gen, Func<IEnumerable<CSharpCodeGenerator>> generatorFactory, StringBuilder buff)
        {
            var generators = generatorFactory.Invoke();
            foreach (var code in generators)
            {
                buff.Append(code.GenerateCode(gen));
            }
            return buff;
        }
    }
}